src/Parser/Template.cs

```diff
@@ -54,6 +57,95 @@ public Template(string template, bool isPositional, List<Literal> literals, List
       Holes = holes.ToArray();
     }
 
+    public string Render(object[] parameters)
+    {
+        var sb = new StringBuilder();
+        int pos = 0;
+        int h = 0;
+        foreach (var literal in Literals)
+        {
+            sb.Append(Value, pos, literal.Print);
+            pos += literal.Print;
+            if (literal.Skip == 0) 
+                pos++;
+            else
+            {
+                pos += literal.Skip;
+                if (IsPositional)
+                    RenderHolePositional(sb, ref Holes[h++], parameters);
+                else
+                    RenderHole(sb, ref Holes[h], parameters[h++]);
+            }
+        }
+        return sb.ToString();
+    }
+
+    private static void RenderHolePositional(StringBuilder sb, ref Hole hole, object[] parameters)
+        => RenderHole(sb, ref hole, parameters[hole.Index], true);
+
+    private static void RenderHole(StringBuilder sb, ref Hole hole, object value, bool legacy = false)
+    {
+        // TODO: handle value == null
+        
+        // TODO: destructuring {@x}
+        
+        if (hole.CaptureType == CaptureType.Stringification)
+        {
+            // TODO: we don't need to support format and alignment here?
+            sb.Append('"').Append(value.ToString()).Append('"');
+            return;
+        }
+
+        // Shortcut common case. It is important to do this before IEnumerable, as string _is_ IEnumerable
+        if (value is string)
+        {
+            AppendValue(sb, ref hole, value, legacy);
+            return;
+        }
+
+        IEnumerable collection;
+        if (!legacy && (collection = value as IEnumerable) != null)
+        {
+            int pos = sb.Length;
+            foreach (var item in collection) {
+                AppendValue(sb, ref hole, item, false);
+                sb.Append(", ");
+            }
+            if (sb.Length > pos) sb.Length -= 2; // Remove trailing ", "
```

Comment:
Agree, we should benchmark it :)

another question, in which case is `sb.Length > pos` false?